### PR TITLE
8324790: ifnode::fold_compares_helper cleanup

### PR DESCRIPTION
**Pull request by caojoshua**
I hope my assumptions  in `filtered_int_type` are correct here:

* we assert that `if_proj` is an `IfTrue` or `IfFalse`, so it is safe to assume `if_proj->_in` is an  `IfNode`
* the 1'th input of a CmpNode is a BoolNode
* Tthe 1'th input of an IfNode is **not always a BoolNode**, it can be a constant. We need to leave this check in.

We also remove a some of the if-checks in `compare_folds_cleanup` which seem unnecessary.

Passes tier1 locally.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8324790](https://bugs.openjdk.org/browse/JDK-8324790): ifnode::fold_compares_helper cleanup (**Enhancement** - P5)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17601/head:pull/17601` \
`$ git checkout pull/17601`

Update a local copy of the PR: \
`$ git checkout pull/17601` \
`$ git pull https://git.openjdk.org/jdk.git pull/17601/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17601`

View PR using the GUI difftool: \
`$ git pr show -t 17601`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17601.diff">https://git.openjdk.org/jdk/pull/17601.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17601#issuecomment-1912828648)

### CODE REVIEW

**Comment by eme64** [comment]
I have not yet looked at your code. But I'm not sure your assumptions are all correct:

> we assert that if_proj is an IfTrue or IfFalse, so it is safe to assume if_proj->_in is an IfNode

Generally, during IGVN, ctrl inputs could always be TOP, if we are in a dead-code place. Are you sure there is always an IfNode above the projections? Maybe that is true in this specific call chain?

> the 1'th input of a CmpNode is a BoolNode

Do you mean the 1st input of BoolNode is a CmpNode? I think we may also fold CmpNodes to constants, or TOP. Is that somehow excluded in your cases?

The problem with testing of this: getting TOP inputs is kinda rare. But you can make it more likely with `StressIGVN`.

**Comment by caojoshua** [comment]
> I have not yet looked at your code. But I'm not sure your assumptions are all correct:
> 
> > we assert that if_proj is an IfTrue or IfFalse, so it is safe to assume if_proj->_in is an IfNode
> 
> Generally, during IGVN, ctrl inputs could always be TOP, if we are in a dead-code place. Are you sure there is always an IfNode above the projections? Maybe that is true in this specific call chain?
> 
> > the 1'th input of a CmpNode is a BoolNode
> 
> Do you mean the 1st input of BoolNode is a CmpNode? I think we may also fold CmpNodes to constants, or TOP. Is that somehow excluded in your cases?
> 
> The problem with testing of this: getting TOP inputs is kinda rare. But you can make it more likely with `StressIGVN`.

I removed all my changes in `filtered_int_type()`. There are some other callers which may have some unexpected constants or TOPs.

My main motivation anyway was in `fold_compares_helper()`, and I think that cleanup is still valid.

File: src/hotspot/share/opto/ifnode.cpp [hunk 1]
```diff
@@ -1015,63 +1015,56 @@ bool IfNode::fold_compares_helper(ProjNode* proj, ProjNode* success, ProjNode* f
...
           return true;
         }
       }
     }
-    lo = nullptr;
-    hi = nullptr;
+    return false;
+  }
+
+  Node* hook = new Node(lo); // Add a use to lo to prevent him from dying
```
Comment (eme64) [inline src/hotspot/share/opto/ifnode.cpp:1029]: Would you mind adding a sanity assert, that both `lo` and `hi` are not `nullptr`?

**Comment by eme64** [comment]
I think this generally looks reasonable. I'm running some testing now...

/reviewers 2

**Review by chhagedorn — APPROVED** [review]
The current patch looks good to me.

**Review by eme64 — APPROVED** [review]
Thanks for the updates, looks good now :)

**Comment by caojoshua** [comment]
@eme64 could you sponsor?

**Pull request closed without merging**